engine/internal/runtime/manager.go

```diff
@@ -0,0 +1,214 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const managerName = "inference-engine"
+
+// NewManager creates a new runtime manager
+func NewManager(k8sClient client.Client, rtClient Client) *Manager {
+	return &Manager{
+		k8sClient:       k8sClient,
+		rtClient:        rtClient,
+		readyRuntimes:   make(map[string]runtime),
+		pendingRuntimes: make(map[string]chan struct{}),
+	}
+}
+
+// Manager manages runtimes.
+type Manager struct {
+	k8sClient client.Client
+	rtClient  Client
+
+	readyRuntimes   map[string]runtime
+	pendingRuntimes map[string]chan struct{}
+	mu              sync.RWMutex
+}
+
+type runtime struct {
+	address string
+}
+
+func (m *Manager) deleteRuntime(modelID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	delete(m.readyRuntimes, modelID)
+	delete(m.pendingRuntimes, modelID)
+}
+
+func (m *Manager) markRuntimeReady(modelID, address string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if r, ok := m.pendingRuntimes[modelID]; ok {
+		close(r)
+		delete(m.pendingRuntimes, modelID)
+		m.readyRuntimes[modelID] = runtime{address: address}
+	}
+}
+
+func (m *Manager) isReady(modelID string) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	_, ok := m.readyRuntimes[modelID]
+	return ok
+}
+
+func (m *Manager) isPending(modelID string) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	_, ok := m.pendingRuntimes[modelID]
+	return ok
+}
+
+// GetLLMAddress returns the address of the LLM.
+func (m *Manager) GetLLMAddress(modelID string) string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if r, ok := m.readyRuntimes[modelID]; ok {
+		return r.address
+	}
+	return ""
+}
+
+// ListSyncedModelIDs returns the list of models that are synced.
+func (m *Manager) ListSyncedModelIDs(ctx context.Context) []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.readyRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// ListInProgressModels returns the list of models that are in progress.
+func (m *Manager) ListInProgressModels() []string {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var modelIDs []string
+	for id := range m.pendingRuntimes {
+		modelIDs = append(modelIDs, id)
+	}
+	return modelIDs
+}
+
+// PullModel pulls the model from the model manager.
+func (m *Manager) PullModel(ctx context.Context, modelID string) error {
+	if m.isReady(modelID) {
```

Comment:
I think there can be a race condition when multiple goroutines run `PullModel` concurrently or when `PullModel` and `Reconcile` interfere badly. Here is an example case:

1. Goroutine A calls `PullModel`. The goroutine starts pulling the model.
2. Goroutine B calls `PullModel` and calls `m.IsReady`. The function returns false.
3. Before B acquires the lock again, goroutine A finishes pulling the model. `Reconcile` calls `m.markRuntimeReady()`.
4. B checks `m.pendingRuntimes` and doesn't find a pending runtime. It calls `DeployRuntime` again.

We can change the code to continue keeping the lock.

```go
func (m *Manager) PullModel(ctx context.Context, modelID string) error {
	m.mu.Lock()
	if _, ok := m.readyRuntimes[modelID]; ok {
		m.mu.Unlock()
		return nil
	}

	var done chan struct{}
	if ch, ok := m.pendingRuntimes[modelID]; ok {
		m.mu.Unlock()
		done = ch
	} else {
		done = make(chan struct{})
		m.pendingRuntimes[modelID] = done
		m.mu.Unlock()
...
```